# MATH2319 Machine Learning
Semester 1, 2020

Assignment 1

*Servet KOCAK* - 3773044

## Question 1

It's known that data preprocessing is a critical component in machine learning and its importance cannot be overstated. During the first question;
* Checking for the missing values
* Imputations with Modes and Medians
* Discretizing Numeric Feature
* Integer Encoding 
* One-Hot Encoding
* Scaling

will be captured.


## Data Description

This dataset concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data.

External link to dataset website [here](https://archive.ics.uci.edu/ml/datasets/Credit+Approval)


The descriptive features below are produced from the `crx.names` file: 

    A1:	b, a.
    A2:	continuous.
    A3:	continuous.
    A4:	u, y, l, t.
    A5:	g, p, gg.
    A6:	c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
    A7:	v, h, bb, j, n, z, dd, ff, o.
    A8:	continuous.
    A9:	t, f.
    A10:	t, f.
    A11:	continuous.
    A12:	t, f.
    A13:	g, p, s.
    A14:	continuous.
    A15:	continuous.
    A16: +,-         (class attribute)

The first step is importing the important libraries; **pandas** and **numpy**.

In [3]:
import pandas as pd
import numpy as np

After importing the libraries, reading the data file follows.

In [4]:
CreditData = pd.read_csv('C:/Users/arnav/OneDrive/Desktop/Machine Learning/Assignment1/Assignment1_Q1_crx.data', sep=",", header=None)
# Definin the column names:
CreditData.columns = ["A1", "A2", "A3", "A4","A5", "A6", "A7", "A8","A9", "A10", "A11", "A12","A13", "A14", "A15", "A16"]

To check if data has been imported correctly, head() function helps. 

In [5]:
CreditData.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

Also, having a look at 10 randomly selected rows in this raw dataset helps to check the dataset as well.

In [6]:
CreditData.sample(n=10, random_state=8)

A1     A2     A3 A4 A5  A6  A7     A8 A9 A10  A11 A12 A13    A14   A15 A16
547  b  23.92  1.500  u  g   d   h  1.875  t   t    6   f   g  00200   327   +
350  a  26.17  2.000  u  g   j   j  0.000  f   f    0   t   g  00276     1   -
18   b  21.83  0.250  u  g   d   h  0.665  t   f    0   t   g  00000     0   +
683  b  36.42  0.750  y  p   d   v  0.585  f   f    0   f   g  00240     3   -
436  b  19.58  0.585  u  g  ff  ff  0.000  f   t    3   f   g  00350   769   -
589  b  25.33  0.580  u  g   c   v  0.290  t   t    7   t   g  00096  5124   +
188  a  28.67  1.040  u  g   c   v  2.500  t   t    5   t   g  00300  1430   +
51   b  26.00  1.000  u  g   q   v  1.750  t   f    0   t   g  00280     0   +
557  a  32.17  1.460  u  g   w   v  1.085  t   t   16   f   g  00120  2079   +
484  b  34.67  1.080  u  g   m   v  1.165  f   f    0   f   s  00028     0   -

To have better understanding of data, the shape of the data can be checked as follows:

In [7]:
CreditData.shape

(690, 16)

As it's seen from the `crx.names` file, **Number of Instances: 690** and **Number of Attributes: 15 + class attribute**. So that's means the dataset imported correctly.

To make sure the *feature types* match the descriptions outlined in the Data Description section, **dtypes** function can be applied.

In [8]:
CreditData.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

## Checking for Missing Values

To check for Missing Values, counting the data is one of the essential steps. 

In [9]:
CreditData.count()

A1     690
A2     690
A3     690
A4     690
A5     690
A6     690
A7     690
A8     690
A9     690
A10    690
A11    690
A12    690
A13    690
A14    690
A15    690
A16    690
dtype: int64

At it seen there is no Null values in the dataset. Also we can check for the Missing Values with following code. 

In [10]:
CreditData.isna().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

Also with counting the Null values is one of the other option but from this output it can be understand that Null values is encoded as other kind of value.

In the next step, the frequency distribution of unique values of A1 and A2 (one categorical, one numerical) columns might help to understand categories.

In [11]:
CreditData['A1'].value_counts()

b    468
a    210
?     12
Name: A1, dtype: int64

In [12]:
CreditData['A2'].value_counts().head()

?        12
22.67     9
20.42     7
19.17     6
25.00     6
Name: A2, dtype: int64

As it seen above there are 12 values are seen as '?' in the in both of the columns. All unusual values must be assigned as NaN's. To do so;

In [13]:
CreditData1 = CreditData.replace('?', np.nan)

With converting "?" to NaN, we can easialy see the Null values in the dataset

In [14]:
CreditData1.isna().sum()

A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

As it understood from the output, in total there are 67 missing values.

The Missing Attribute Values below are produced from the `crx.names` file: 

    37 cases (5%) have one or more missing values.  The missing values from particular attributes are:

    A1:  12
    A2:  12
    A4:   6
    A5:   6
    A6:   9
    A7:   9
    A14: 13

With the code above, the Missing Attribute Values is confirmed with the output.

## Imputing the Missing Values

After setting all outliers/ unusual values to missing values, In this step, any missing values  will be imputed with
* the **mode** for categorical features 
* the **median** for numerical features.

### Numerical Features

Since, `A2` and `A14` are numerical attributes which have Null values, the missing values need to be imputed with the mean for each column. Firstly, all the columns median values can be checked as below;

In [15]:
CreditData1[['A2','A14']].median().round(2)

A2      28.46
A14    160.00
dtype: float64

These values will be used to impute the missing values.

In [16]:
CreditData1['A2'] = CreditData1['A2'].apply(pd.to_numeric)

In [17]:
CreditData1['A2'].median() # This value will be used to impute the missing values in A2 column.

28.46

In [18]:
CreditData1['A2'] = CreditData1['A2'].fillna(CreditData1['A2'].median())

In [19]:
CreditData1['A2'].isna().sum()

0

As it seen from above, there is no missing data left in the A2 column after imputation. 

In [20]:
CreditData1['A14'] = CreditData1['A14'].apply(pd.to_numeric)
print('This value ' + '"'+ str(CreditData1['A14'].median())+'"'+ ' will be used to impute the missing values for A14 column.')
CreditData1['A14'] = CreditData1['A14'].fillna(CreditData1['A14'].median())
print('After imputation, there is ' + str(CreditData1['A14'].isna().sum())+ ' missing values left.')

This value "160.0" will be used to impute the missing values for A14 column.
After imputation, there is 0 missing values left.


### Categorical Features

Since, `A1`,`A4`,`A5`,`A6` and`A7` are categorical attributes which have Null values, the missing values need to be imputed with the mode for each column. Firstly, all the columns median values can be checked as below;

In [21]:
CreditData1[['A1','A4','A5','A6','A7']].mode()

A1 A4 A5 A6 A7
0  b  u  g  c  v

In [22]:
CreditData1['A1'] = CreditData1['A1'].fillna('b') # imputing the missing values with 'b'
print(CreditData1['A1'].isna().sum())

0


As it seen above there is no missing values left after imputation.

In [23]:
CreditData1['A4'] = CreditData1['A4'].fillna('u')
print(CreditData1['A4'].isna().sum())

0


In [24]:
CreditData1['A5'] = CreditData1['A5'].fillna('g')
print(CreditData1['A5'].isna().sum())

0


In [25]:
CreditData1['A6'] = CreditData1['A6'].fillna('c')
print(CreditData1['A6'].isna().sum())

0


In [26]:
CreditData1['A7'] = CreditData1['A7'].fillna('v')
print(CreditData1['A7'].isna().sum())

0


To confirm one more time to see there is no missing values left;

In [27]:
CreditData1.isna().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

## Discretizing Numeric Features 

For the `A2` numerical descriptive feature, it will be discretized via equal-frequency binning with 3 bins named *"low"*, *"medium",* and *"high"*, and then integer encoding will be used.

> The **qcut method** is used in the Pandas module, which performs quantile-based discretization.

In [28]:
df_cat = CreditData1.copy()
df_cat['A2'] = df_cat['A2']

df_cat['A2'] = pd.qcut(df_cat['A2'], q=3, 
                                     labels=['low', 'medium', 'high'])


To make sure the dicretization is performed correctly using the value_counts method in Pandas.

In [29]:
df_cat['A2'].value_counts()

medium    231
low       230
high      229
Name: A2, dtype: int64

### Integer-Encoding

A nominal descriptive feature always needs be encoded using one-hot-encoding. However, an ordinal descriptive feature can be encoded via either one-hot-encoding or integer-encoding. For the latter, we can use the replace() function in Pandas. Let's do a simple example. The "mean_area" feature we defined above can in fact be considered to be "ordinal" as there is a natural ordering between the levels of small, average, and large. Let's encode this feature using integer-encoding where each level corresponds to the integers 0, 1, and 2 respectively. 

In [30]:
# Before using the replace() function, the mapping between the levels and the integers needs to be defined.
level_mapping = {'low': 0, 'medium': 1, 'high': 2} 
df_cat_integer = df_cat.copy()
df_cat_integer['A2'] = df_cat_integer['A2'].replace(level_mapping)
df_cat_integer.head(5)

A1  A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b   1  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a   2  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a   1  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b   1  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b   0  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +

## One-Hot-Encoding


Nominal descriptive features must always be encoded using one-hot-encoding. We can use the get_dummies() method in the Pandas module for one-hot-encoding. This is a very simple and effective method as you can do one-hot-encoding for multiple features at once and you can even supply a custom prefix for each categorical feature. If you omit the columns parameter, the **get_dummies()** method will intelligently do one-hot-encoding for all features that are not numeric. Thus, get_dummies() can be used in "automatic" mode.

Get a list of all descriptive categorical features. Display value counts for each one of these features. Comment on which feature(s) appear to be ordinal.

In [31]:
# Since it's known that A16 is target, this column will be droped here and encoded in the next sections.
Data_Cat = df_cat_integer.drop(columns = 'A16') 
# Rest of the Data, will be encoded in this section.
categorical_cols = Data_Cat.columns[Data_Cat.dtypes == np.object].tolist()
categorical_cols

['A1', 'A4', 'A5', 'A6', 'A7', 'A9', 'A10', 'A12', 'A13']

So all the categorical features appear to be nominal. The one-hot encoding for all the descriptive categorical features will be performed and  this encoded data frame will be called as *Data_encoded*. 

If a categorical descriptive feature has **only 2 levels**, it will be encoded with only **one binary variable**. 

For other categorical features (*with more than 2 levels*), **regular one-hot-encoding** will be used (where number of binary variables are equal to the number of distinct levels).


In [34]:
Data_encoded = Data_Cat.copy()  # retain original Data without encoding for further analysis

# if a categorical descriptive feature has only 2 levels,
# define only one binary variable
for col in categorical_cols:
    q = len(Data_encoded[col].unique())
    if (q == 2):
        Data_encoded[col] = pd.get_dummies(Data_encoded[col], drop_first=True)

# for other categorical features (with > 2 levels), 
# perform regular one-hot-encoding using pd.get_dummies()
# if a feature is numeric, it will be untouched
Data_encoded = pd.get_dummies(Data_encoded)

In [35]:
print(f'The Shape of Data_encoded is {Data_encoded.shape}\n')  

Data_encoded.sample(n=4, random_state=11)

The Shape of Data_encoded is (690, 42)



A1  A2      A3    A8  A9  A10  A11  A12    A14  A15  ...  A7_ff  A7_h  \
214   1   1   2.710  5.25   1    1    1    0  211.0    0  ...      0     0   
545   1   2  11.000  1.50   1    0    0    0    0.0    0  ...      0     0   
436   1   0   0.585  0.00   0    1    3    0  350.0  769  ...      1     0   
201   0   2   1.000  2.25   1    0    0    1    0.0  300  ...      0     0   

     A7_j  A7_n  A7_o  A7_v  A7_z  A13_g  A13_p  A13_s  
214     0     0     0     1     0      1      0      0  
545     0     0     0     1     0      0      0      1  
436     0     0     0     0     0      1      0      0  
201     0     0     0     0     0      1      0      0  

[4 rows x 42 columns]

After encoding all the categorical features, we end up with a data frame that is all numerical. Get a description of Data_encoded with `include='all'` option.

In [108]:
Data_encoded.describe(include='all').round(3)

A1       A2       A3       A8       A9      A10      A11      A12  \
count  690.000  690.000  690.000  690.000  690.000  690.000  690.000  690.000   
mean     0.696    0.999    4.759    2.223    0.523    0.428    2.400    0.458   
std      0.460    0.816    4.978    3.347    0.500    0.495    4.863    0.499   
min      0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000   
25%      0.000    0.000    1.000    0.165    0.000    0.000    0.000    0.000   
50%      1.000    1.000    2.750    1.000    1.000    0.000    0.000    0.000   
75%      1.000    2.000    7.208    2.625    1.000    1.000    3.000    1.000   
max      1.000    2.000   28.000   28.500    1.000    1.000   67.000    1.000   

            A14         A15   ...       A7_ff   A7_h     A7_j     A7_n  \
count   690.000     690.000   ...     690.000  690.0  690.000  690.000   
mean    183.562    1017.386   ...       0.083    0.2    0.012    0.006   
std     172.190    5210.103   ...       0.275    0.4    0.107    0.076   
min       0.000       0.000   ...       0.000    0.0    0.000    0.000   
25%      80.000       0.000   ...       0.000    0.0    0.000    0.000   
50%     160.000       5.000   ...       0.000    0.0    0.000    0.000   
75%     272.000     395.500   ...       0.000    0.0    0.000    0.000   
max    2000.000  100000.000   ...       1.000    1.0    1.000    1.000   

          A7_o     A7_v     A7_z    A13_g    A13_p    A13_s  
count  690.000  690.000  690.000  690.000  690.000  690.000  
mean     0.003    0.591    0.012    0.906    0.012    0.083  
std      0.054    0.492    0.107    0.292    0.107    0.275  
min      0.000    0.000    0.000    0.000    0.000    0.000  
25%      0.000    0.000    0.000    1.000    0.000    0.000  
50%      0.000    1.000    0.000    1.000    0.000    0.000  
75%      0.000    1.000    0.000    1.000    0.000    0.000  
max      1.000    1.000    1.000    1.000    1.000    1.000  

[8 rows x 42 columns]

### Encoding The Target Feature 

As it in the One-Hot Encoding section, A16 was excluded from the whole dataset. In this section A16 column will be encoding using `.replace` command. 

In [33]:
A16 = df_cat_integer['A16']
np.unique(A16, return_counts=True)

(array(['+', '-'], dtype=object), array([307, 383]))

The target feature will be label-encoded since its known that target column is binary, the positive class is encoded as 1.

In [120]:
A16 = A16.replace({'+': 1 , '-':0})
A16.value_counts()

TypeError: Cannot compare types 'ndarray(dtype=int64)' and 'str'

## Scaling Descriptive Features

Once all categorical descriptive features are encoded, all features in this transformed dataset will be numerical. It is always a good idea to scale these numerical descriptive features before fitting any models, as scaling is mandatory for some important class of models such as nearest neighbors, SVMs, and deep learning.

Scaling of each descriptive feature is done via standardization. That is, each value of the descriptive feature is scaled by removing the mean and dividing by the standard deviation of that feature. `This ensures that, after scaling, each descriptive feature has a 0 mean and 1 standard deviation. ` 

$$ Standart Scaling =\frac{value - mean}{standart deviation}$$



Before fitting any Scikit-Learn models, any Pandas series or data frame is converted to a NumPy array using the `.values` method in Pandas.

In [127]:
from sklearn import preprocessing

Standard= pd.DataFrame(preprocessing.StandardScaler().fit_transform(Data_encoded.values), columns=Data_encoded.columns)

Standard.head()

A1        A2        A3        A8       A9       A10       A11  \
0  0.661438  0.001777 -0.956613 -0.291083  0.95465  1.157144 -0.288101   
1 -1.511858  1.227857 -0.060051  0.244190  0.95465  1.157144  0.740830   
2 -1.511858  0.001777 -0.856102 -0.216324  0.95465 -0.864196 -0.493887   
3  0.661438  0.001777 -0.647038  0.456505  0.95465  1.157144  0.535044   
4  0.661438 -1.224303  0.174141 -0.153526  0.95465 -0.864196 -0.493887   

        A12       A14       A15    ...        A7_ff  A7_h      A7_j     A7_n  \
0 -0.919195  0.107155 -0.195413    ...    -0.300079  -0.5 -0.108306 -0.07636   
1 -0.919195 -0.816912 -0.087852    ...    -0.300079   2.0 -0.108306 -0.07636   
2 -0.919195  0.560471 -0.037144    ...    -0.300079   2.0 -0.108306 -0.07636   
3  1.087908 -0.485643 -0.194837    ...    -0.300079  -0.5 -0.108306 -0.07636   
4 -0.919195 -0.369408 -0.195413    ...    -0.300079  -0.5 -0.108306 -0.07636   

       A7_o      A7_v      A7_z     A13_g     A13_p     A13_s  
0 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
1 -0.053916 -1.202834 -0.108306  0.322490 -0.108306 -0.300079  
2 -0.053916 -1.202834 -0.108306  0.322490 -0.108306 -0.300079  
3 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
4 -0.053916  0.831370 -0.108306 -3.100868 -0.108306  3.332456  

[5 rows x 42 columns]

Before calling the last version of the dataset as a clean dataset, the target column needs to be combined with the last dataset. Since A16 is known as a target variable, it needs to be the last column in the clean data and its name needs to be target.

In [128]:
With_Target= pd.concat([Standard,A16], axis=1) # Combining part
df_target =With_Target.rename(columns= {'A16':'Target'} )
df_target.head()

A1        A2        A3        A8       A9       A10       A11  \
0  0.661438  0.001777 -0.956613 -0.291083  0.95465  1.157144 -0.288101   
1 -1.511858  1.227857 -0.060051  0.244190  0.95465  1.157144  0.740830   
2 -1.511858  0.001777 -0.856102 -0.216324  0.95465 -0.864196 -0.493887   
3  0.661438  0.001777 -0.647038  0.456505  0.95465  1.157144  0.535044   
4  0.661438 -1.224303  0.174141 -0.153526  0.95465 -0.864196 -0.493887   

        A12       A14       A15   ...    A7_h      A7_j     A7_n      A7_o  \
0 -0.919195  0.107155 -0.195413   ...    -0.5 -0.108306 -0.07636 -0.053916   
1 -0.919195 -0.816912 -0.087852   ...     2.0 -0.108306 -0.07636 -0.053916   
2 -0.919195  0.560471 -0.037144   ...     2.0 -0.108306 -0.07636 -0.053916   
3  1.087908 -0.485643 -0.194837   ...    -0.5 -0.108306 -0.07636 -0.053916   
4 -0.919195 -0.369408 -0.195413   ...    -0.5 -0.108306 -0.07636 -0.053916   

       A7_v      A7_z     A13_g     A13_p     A13_s  Target  
0  0.831370 -0.108306  0.322490 -0.108306 -0.300079       1  
1 -1.202834 -0.108306  0.322490 -0.108306 -0.300079       1  
2 -1.202834 -0.108306  0.322490 -0.108306 -0.300079       1  
3  0.831370 -0.108306  0.322490 -0.108306 -0.300079       1  
4  0.831370 -0.108306 -3.100868 -0.108306  3.332456       1  

[5 rows x 43 columns]

Once the data is clean, the final dataset can be called as df_clean.

In [129]:
df_clean = df_target
df_clean.shape

(690, 43)

As it seen above, the clean dataset has 690 rows ans 43 columns.

To check the all the variable's descriptive statistics;

In [130]:
df_clean.describe(include='all').round(3)

A1       A2       A3       A8       A9      A10      A11      A12  \
count  690.000  690.000  690.000  690.000  690.000  690.000  690.000  690.000   
mean     0.000   -0.000    0.000    0.000    0.000   -0.000    0.000   -0.000   
std      1.001    1.001    1.001    1.001    1.001    1.001    1.001    1.001   
min     -1.512   -1.224   -0.957   -0.665   -1.048   -0.864   -0.494   -0.919   
25%     -1.512   -1.224   -0.756   -0.616   -1.048   -0.864   -0.494   -0.919   
50%      0.661    0.002   -0.404   -0.366    0.955   -0.864   -0.494   -0.919   
75%      0.661    1.228    0.492    0.120    0.955    1.157    0.123    1.088   
max      0.661    1.228    4.672    7.858    0.955    1.157   13.294    1.088   

           A14      A15   ...        A7_h     A7_j     A7_n     A7_o     A7_v  \
count  690.000  690.000   ...     690.000  690.000  690.000  690.000  690.000   
mean     0.000   -0.000   ...       0.000    0.000   -0.000   -0.000   -0.000   
std      1.001    1.001   ...       1.001    1.001    1.001    1.001    1.001   
min     -1.067   -0.195   ...      -0.500   -0.108   -0.076   -0.054   -1.203   
25%     -0.602   -0.195   ...      -0.500   -0.108   -0.076   -0.054   -1.203   
50%     -0.137   -0.194   ...      -0.500   -0.108   -0.076   -0.054    0.831   
75%      0.514   -0.119   ...      -0.500   -0.108   -0.076   -0.054    0.831   
max     10.557   19.012   ...       2.000    9.233   13.096   18.547    0.831   

          A7_z    A13_g    A13_p    A13_s   Target  
count  690.000  690.000  690.000  690.000  690.000  
mean    -0.000    0.000   -0.000   -0.000    0.445  
std      1.001    1.001    1.001    1.001    0.497  
min     -0.108   -3.101   -0.108   -0.300    0.000  
25%     -0.108    0.322   -0.108   -0.300    0.000  
50%     -0.108    0.322   -0.108   -0.300    0.000  
75%     -0.108    0.322   -0.108   -0.300    1.000  
max      9.233    0.322    9.233    3.332    1.000  

[8 rows x 43 columns]

To check the first 5 rows in the clean dataset;

In [131]:
df_clean.head(5)

A1        A2        A3        A8       A9       A10       A11  \
0  0.661438  0.001777 -0.956613 -0.291083  0.95465  1.157144 -0.288101   
1 -1.511858  1.227857 -0.060051  0.244190  0.95465  1.157144  0.740830   
2 -1.511858  0.001777 -0.856102 -0.216324  0.95465 -0.864196 -0.493887   
3  0.661438  0.001777 -0.647038  0.456505  0.95465  1.157144  0.535044   
4  0.661438 -1.224303  0.174141 -0.153526  0.95465 -0.864196 -0.493887   

        A12       A14       A15   ...    A7_h      A7_j     A7_n      A7_o  \
0 -0.919195  0.107155 -0.195413   ...    -0.5 -0.108306 -0.07636 -0.053916   
1 -0.919195 -0.816912 -0.087852   ...     2.0 -0.108306 -0.07636 -0.053916   
2 -0.919195  0.560471 -0.037144   ...     2.0 -0.108306 -0.07636 -0.053916   
3  1.087908 -0.485643 -0.194837   ...    -0.5 -0.108306 -0.07636 -0.053916   
4 -0.919195 -0.369408 -0.195413   ...    -0.5 -0.108306 -0.07636 -0.053916   

       A7_v      A7_z     A13_g     A13_p     A13_s  Target  
0  0.831370 -0.108306  0.322490 -0.108306 -0.300079       1  
1 -1.202834 -0.108306  0.322490 -0.108306 -0.300079       1  
2 -1.202834 -0.108306  0.322490 -0.108306 -0.300079       1  
3  0.831370 -0.108306  0.322490 -0.108306 -0.300079       1  
4  0.831370 -0.108306 -3.100868 -0.108306  3.332456       1  

[5 rows x 43 columns]

To export the clean dataset;

In [132]:
df_clean.to_csv('df_clean.csv', index=False)

---